src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitSynchronizationContext.cs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    [DebuggerDisplay("{_state,nq}")]
+    internal class CircuitSynchronizationContext : SynchronizationContext
+    {
+        private static readonly ContextCallback ExecutionContextThunk = (object state) =>
+        {
+            var item = (WorkItem)state;
+            item.Callback(item.State);
+        };
+
+        private static readonly WaitCallback BackgroundWorkThunk = (object state) =>
+        {
+            var context = (CircuitSynchronizationContext)state;
+            context.ExecuteBackgroundWork();
+        };
+
+        private State _state;
+
+        public event UnhandledExceptionEventHandler UnhandledException;
```

Comment:
I chose to make this pump unhandled exceptions on the async call path via an event. The idea is that whatever error handling strategy we need can be attached to this. 

Note that the typical call paths here will use `InvokeAsync` and so will get their errors in a normal way 😆 

You'd only get an unhandled exception if you threw from an async void method.